From 6dd118db2beba52a2e74e6005af1b68144050834 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 17 Feb 2014 00:05:26 -0500
Subject: [PATCH] Chat API

Initial for Fanciful Credits to http://forums.bukkit.org/threads/lib-fanciful-pleasant-chat-message-formatting.195148/
But did a lot of bug fixing and improving upon it, and designing into the Dragon API
---
 .../paperdragon/api/CraftDragonAPI.java       |   2 +
 .../paperdragon/api/FancifulChatAPIImpl.java  | 282 ++++++++++++++++++
 .../paperdragon/api/chat/CraftDAPI_Chat.java  |  40 +++
 .../net/minecraft/server/ChatHoverable.java   |  28 +-
 .../java/net/minecraft/server/Entity.java     |   1 +
 5 files changed, 339 insertions(+), 14 deletions(-)
 create mode 100644 src/main/java/com/domnian/paperdragon/api/FancifulChatAPIImpl.java
 create mode 100644 src/main/java/com/domnian/paperdragon/api/chat/CraftDAPI_Chat.java

diff --git a/src/main/java/com/domnian/paperdragon/api/CraftDragonAPI.java b/src/main/java/com/domnian/paperdragon/api/CraftDragonAPI.java
index b11af17ef..81ceb7f59 100644
--- a/src/main/java/com/domnian/paperdragon/api/CraftDragonAPI.java
+++ b/src/main/java/com/domnian/paperdragon/api/CraftDragonAPI.java
@@ -23,6 +23,7 @@
 
 package com.domnian.paperdragon.api;
 
+import com.domnian.paperdragon.api.chat.CraftDAPI_Chat;
 import com.domnian.paperdragon.api.meta.CraftDAPI_Meta;
 
 public final class CraftDragonAPI extends DragonAPI {
@@ -34,6 +35,7 @@ public final class CraftDragonAPI extends DragonAPI {
         entity = new CraftDAPI_Entity();
         misc = new CraftDAPI_Misc();
         meta = new CraftDAPI_Meta();
+        chat = new CraftDAPI_Chat();
     }
 
 }
diff --git a/src/main/java/com/domnian/paperdragon/api/FancifulChatAPIImpl.java b/src/main/java/com/domnian/paperdragon/api/FancifulChatAPIImpl.java
new file mode 100644
index 000000000..dadaaaee8
--- /dev/null
+++ b/src/main/java/com/domnian/paperdragon/api/FancifulChatAPIImpl.java
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2016 Starlis LLC / Daniel Ennis (Aikar) - MIT License
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining
+ *  a copy of this software and associated documentation files (the
+ *  "Software"), to deal in the Software without restriction, including
+ *  without limitation the rights to use, copy, modify, merge, publish,
+ *  distribute, sublicense, and/or sell copies of the Software, and to
+ *  permit persons to whom the Software is furnished to do so, subject to
+ *  the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package com.domnian.paperdragon.api;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.domnian.paperdragon.api.HiddenItemMeta;
+import net.minecraft.server.*;
+
+import net.minecraft.server.PacketPlayOutTitle.EnumTitleAction;
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.json.JSONException;
+import org.json.JSONStringer;
+
+/**
+ * Credits to original version @
+ * http://forums.bukkit.org/threads/lib-fanciful-pleasant-chat-message-formatting.195148/
+ *
+ * Improved to use proper components to fix issues
+ * @author aikar
+ *
+ */
+public class FancifulChatAPIImpl implements DAPI_Chat.FancifulChat {
+    private final List<MessagePart> messageParts = new ArrayList<>();
+    private ChatComponentText component;
+    private boolean isDirty = true;
+
+    public FancifulChatAPIImpl file(final String path) {
+        onClick("open_file", path);
+        return this;
+    }
+
+    public FancifulChatAPIImpl link(final String url) {
+        onClick("open_url", url);
+        return this;
+    }
+
+    public FancifulChatAPIImpl suggest(final String command) {
+        onClick("suggest_command", command);
+        return this;
+    }
+
+    public FancifulChatAPIImpl command(final String command) {
+        onClick("run_command", command);
+        return this;
+    }
+
+    public FancifulChatAPIImpl advancementTooltip(final String name) {
+        //TODO: onHover("show_advancement", "advancement." + name);
+        return this;
+    }
+
+    public FancifulChatAPIImpl page(String page) {
+        onClick("change_page", page);
+        return this;
+    }
+
+    public FancifulChatAPIImpl copy(String text) {
+        return null;
+    }
+
+    public FancifulChatAPIImpl itemTooltip(final String itemJSON) {
+        //TODO: onHover("show_item", itemJSON);
+        return this;
+    }
+
+    public FancifulChatAPIImpl itemTooltip(final ItemStack itemStack) {
+        onHoverItem(itemStack);
+        return this;
+    }
+
+    public FancifulChatAPIImpl tooltip(final String text) {
+        final String[] lines = text.split("\\n");
+        if (lines.length <= 1) {
+            onHoverText(text);
+        } else {
+            itemTooltip(makeMultilineTooltip(lines));
+        }
+        return this;
+    }
+
+    public FancifulChatAPIImpl entityTooltip(Entity entity) {
+        onHoverEntity(entity);
+        return this;
+    }
+
+    public FancifulChatAPIImpl then(final Object obj) {
+        messageParts.add(new MessagePart(obj.toString()));
+        isDirty = true;
+        return this;
+    }
+
+    public void send(Player player) {
+        sendPacket(player, new PacketPlayOutChat(buildComponent(), net.minecraft.server.ChatMessageType.SYSTEM, SystemUtils.getNullUUID()));
+    }
+
+    public static void sendTitle(Player player, FancifulChatAPIImpl title, FancifulChatAPIImpl subtitle, Integer fadeIn, Integer stay, Integer fadeOut) {
+        if (fadeIn != null && stay != null && fadeOut != null) {
+            sendPacket(player, new PacketPlayOutTitle(fadeIn, stay, fadeOut));
+        }
+        if (title != null) {
+            sendPacket(player, new PacketPlayOutTitle(EnumTitleAction.TITLE, title.buildComponent()));
+        }
+        if (subtitle != null) {
+            sendPacket(player, new PacketPlayOutTitle(EnumTitleAction.SUBTITLE, subtitle.buildComponent()));
+        }
+    }
+
+    private static void sendPacket(Player player, Packet<?> packet) {
+        final PlayerConnection conn = ((CraftPlayer) player).getHandle().playerConnection;
+        if (conn == null) {
+            return;
+        }
+        conn.sendPacket(packet);
+    }
+
+    private ChatComponentText buildComponent() {
+        if (component != null && !isDirty) {
+            return component;
+        }
+        isDirty = false;
+        component = new ChatComponentText("");
+        for (MessagePart msg : messageParts) {
+            for (IChatBaseComponent cmp : msg.getComponents()) {
+                component.addSibling(cmp);
+            }
+        }
+        return component;
+    }
+
+    protected MessagePart latest() {
+        return messageParts.get(messageParts.size() - 1);
+    }
+
+    protected String makeMultilineTooltip(final String[] lines) {
+        final JSONStringer json = new JSONStringer();
+        try {
+            json.object().key("id").value(1);
+            json.key("tag").object().key("display").object();
+            json.key("Name").value("\\u00A7f" + lines[0].replace("\"", "\\\""));
+            json.key("Lore").array();
+            for (int i = 1; i < lines.length; i++) {
+                json.value(lines[i].isEmpty() ? " " : lines[i].replace("\"", "\\\""));
+            }
+            json.endArray().endObject().endObject().endObject();
+        } catch (final JSONException e) {
+            throw new RuntimeException("invalid tooltip");
+        }
+        return json.toString();
+    }
+
+    protected void onClick(final String name, final String data) {
+        final MessagePart latest = latest();
+        latest.clickActionName = name;
+        latest.clickActionData = data;
+        isDirty = true;
+    }
+
+    protected void onHoverText(final String data) {
+        final MessagePart latest = latest();
+        latest.hoverActionName = "show_text";
+        latest.hoverTextData = data;
+        isDirty = true;
+    }
+
+    protected void onHoverItem(ItemStack itemStack) {
+        final MessagePart latest = latest();
+        latest.hoverActionName = "show_item";
+        latest.hoverItemData = CraftItemStack.asNMSCopy(itemStack);
+        isDirty = true;
+    }
+
+    private void onHoverEntity(Entity entity) {
+        final MessagePart latest = latest();
+        latest.hoverActionName = "show_entity";
+        latest.hoverEntityData = ((CraftEntity)entity).getHandle();
+        isDirty = true;
+    }
+
+    static final class MessagePart {
+
+
+        String clickActionName = null, clickActionData = null,
+                hoverActionName = null, hoverTextData = null;
+        net.minecraft.server.ItemStack hoverItemData = null;
+        net.minecraft.server.Entity hoverEntityData = null;
+        final IChatBaseComponent[] components;
+
+        MessagePart(final String text) {
+            components = CraftChatMessage.fromString(text, true);
+        }
+
+        public IChatBaseComponent[] getComponents() {
+            for (IChatBaseComponent component : components) {
+                if (clickActionName != null) {
+                    switch (clickActionName) {
+                        case "open_file":
+                            component.getChatModifier().setChatClickable(
+                                    new ChatClickable(ChatClickable.EnumClickAction.OPEN_FILE, clickActionData));
+                            break;
+                        case "open_url":
+                            component.getChatModifier().setChatClickable(
+                                    new ChatClickable(ChatClickable.EnumClickAction.OPEN_URL, clickActionData));
+                            break;
+                        case "suggest_command":
+                            component.getChatModifier().setChatClickable(
+                                    new ChatClickable(ChatClickable.EnumClickAction.SUGGEST_COMMAND, clickActionData));
+                            break;
+                        case "run_command":
+                            component.getChatModifier().setChatClickable(
+                                    new ChatClickable(ChatClickable.EnumClickAction.RUN_COMMAND, clickActionData));
+                            break;
+                        case "change_page":
+                            component.getChatModifier().setChatClickable(
+                                    new ChatClickable(ChatClickable.EnumClickAction.CHANGE_PAGE, clickActionData));
+                            break;
+                        case "copy_to_clipboard":
+                            component.getChatModifier().setChatClickable(
+                                    new ChatClickable(ChatClickable.EnumClickAction.COPY_TO_CLIPBOARD, clickActionData));
+                    }
+                }
+                if (hoverActionName != null) {
+                    switch (hoverActionName) {
+                        case "show_item":
+                            if (hoverItemData != null) {
+                                NBTTagCompound root = hoverItemData.save(new NBTTagCompound());
+                                NBTTagCompound hover = null;
+                                if (root.hasKey("tag")) {
+                                    hover = HiddenItemMeta.filterItemLore(root.getCompound("tag"), false);
+                                }
+                                component.getChatModifier().setChatHoverable(new ChatHoverable(ChatHoverable.EnumHoverAction.SHOW_ITEM, new ChatHoverable.c(hoverItemData.getItem(), hoverItemData.getCount(), hover)));
+                            }
+                            break;
+                        case "show_text":
+                            if (hoverTextData != null) {
+                                final ChatComponentText hover = new ChatComponentText(hoverTextData);
+                                component.getChatModifier().setChatHoverable(new ChatHoverable(ChatHoverable.EnumHoverAction.SHOW_TEXT, hover));
+                            }
+                            break;
+                        case "show_entity":
+                            if (hoverEntityData != null) {
+                                component.getChatModifier().setChatHoverable(hoverEntityData.createHoverEvent());
+                            }
+                            break;
+                        case "hover_json":
+                            //todo: ChatHoverable.a(JsonObject jsonobject)
+                    }
+                }
+            }
+
+            return components;
+        }
+    }
+}
diff --git a/src/main/java/com/domnian/paperdragon/api/chat/CraftDAPI_Chat.java b/src/main/java/com/domnian/paperdragon/api/chat/CraftDAPI_Chat.java
new file mode 100644
index 000000000..afc5e9e73
--- /dev/null
+++ b/src/main/java/com/domnian/paperdragon/api/chat/CraftDAPI_Chat.java
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2016 Starlis LLC / Daniel Ennis (Aikar) - MIT License
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining
+ *  a copy of this software and associated documentation files (the
+ *  "Software"), to deal in the Software without restriction, including
+ *  without limitation the rights to use, copy, modify, merge, publish,
+ *  distribute, sublicense, and/or sell copies of the Software, and to
+ *  permit persons to whom the Software is furnished to do so, subject to
+ *  the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package com.domnian.paperdragon.api.chat;
+
+import com.domnian.paperdragon.api.DAPI_Chat;
+import com.domnian.paperdragon.api.FancifulChatAPIImpl;
+import org.bukkit.entity.Player;
+
+public class CraftDAPI_Chat implements DAPI_Chat {
+    @Override
+    public FancifulChat createFancy() {
+        return new FancifulChatAPIImpl();
+    }
+
+    @Override
+    public void sendTitle(Player player, FancifulChat title, FancifulChat subtitle, Integer fadeIn, Integer stay, Integer fadeOut) {
+        FancifulChatAPIImpl.sendTitle(player, (FancifulChatAPIImpl) title, (FancifulChatAPIImpl) subtitle, fadeIn, stay, fadeOut);
+    }
+}
diff --git a/src/main/java/net/minecraft/server/ChatHoverable.java b/src/main/java/net/minecraft/server/ChatHoverable.java
index 5e366c413..6f6f8e030 100644
--- a/src/main/java/net/minecraft/server/ChatHoverable.java
+++ b/src/main/java/net/minecraft/server/ChatHoverable.java
@@ -17,21 +17,21 @@ import org.apache.logging.log4j.Logger;
 public class ChatHoverable {
 
     private static final Logger LOGGER = LogManager.getLogger();
-    private final ChatHoverable.EnumHoverAction<?> b;
-    private final Object c;
+    private final ChatHoverable.EnumHoverAction<?> action; // PaperDragon - OBFHELPER - decompile fix
+    private final Object value; // PaperDragon - OBFHELPER - decompile fix
 
     public <T> ChatHoverable(ChatHoverable.EnumHoverAction<T> chathoverable_enumhoveraction, T t0) {
-        this.b = chathoverable_enumhoveraction;
-        this.c = t0;
+        this.action = chathoverable_enumhoveraction; // PaperDragon - decompile fix
+        this.value = t0; // PaperDragon - decompile fix
     }
 
     public ChatHoverable.EnumHoverAction<?> a() {
-        return this.b;
+        return this.action; // PaperDragon - decompile fix
     }
 
     @Nullable
     public <T> T a(ChatHoverable.EnumHoverAction<T> chathoverable_enumhoveraction) {
-        return this.b == chathoverable_enumhoveraction ? chathoverable_enumhoveraction.b(this.c) : null;
+        return this.action == chathoverable_enumhoveraction ? chathoverable_enumhoveraction.b(this.value) : null; // PaperDragon - decompile fix
     }
 
     public boolean equals(Object object) {
@@ -40,20 +40,20 @@ public class ChatHoverable {
         } else if (object != null && this.getClass() == object.getClass()) {
             ChatHoverable chathoverable = (ChatHoverable) object;
 
-            return this.b == chathoverable.b && Objects.equals(this.c, chathoverable.c);
+            return this.action == chathoverable.action && Objects.equals(this.value, chathoverable.value); // PaperDragon - decompile fix
         } else {
             return false;
         }
     }
 
     public String toString() {
-        return "HoverEvent{action=" + this.b + ", value='" + this.c + '\'' + '}';
+        return "HoverEvent{action=" + this.action + ", value='" + this.value + '\'' + '}'; // PaperDragon - decompile fix
     }
 
     public int hashCode() {
-        int i = this.b.hashCode();
+        int i = this.action.hashCode(); // PaperDragon - decompile fix
 
-        i = 31 * i + (this.c != null ? this.c.hashCode() : 0);
+        i = 31 * i + (this.value != null ? this.value.hashCode() : 0); // PaperDragon - decompile fix
         return i;
     }
 
@@ -85,8 +85,8 @@ public class ChatHoverable {
     public JsonObject b() {
         JsonObject jsonobject = new JsonObject();
 
-        jsonobject.addProperty("action", this.b.b());
-        jsonobject.add("contents", this.b.a(this.c));
+        jsonobject.addProperty("action", this.action.b()); // PaperDragon - decompile fix
+        jsonobject.add("contents", this.action.a(this.value)); // PaperDragon - decompile fix
         return jsonobject;
     }
 
@@ -132,7 +132,7 @@ public class ChatHoverable {
         }
 
         private T b(Object object) {
-            return object;
+            return (T) object; // PaperDragon - decompile fix
         }
 
         @Nullable
@@ -165,7 +165,7 @@ public class ChatHoverable {
         @Nullable
         private final NBTTagCompound c;
 
-        c(Item item, int i, @Nullable NBTTagCompound nbttagcompound) {
+        public c(Item item, int i, @Nullable NBTTagCompound nbttagcompound) { // PaperDragon - default -> public
             this.a = item;
             this.b = i;
             this.c = nbttagcompound;
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 598fe80cc..853071dbe 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2910,6 +2910,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
         return this.getDirection();
     }
 
+    public ChatHoverable createHoverEvent() { return this.ca(); } // PaperDragon - OBFHELPER
     protected ChatHoverable ca() {
         return new ChatHoverable(ChatHoverable.EnumHoverAction.SHOW_ENTITY, new ChatHoverable.b(this.getEntityType(), this.getUniqueID(), this.getDisplayName()));
     }
-- 
2.20.1.windows.1

